### PR TITLE
Feature/gh 31/modular loading

### LINKED ISSUE #31: [FEATURE] Allow modular model saving/loading
## Summary
With modular model saving/loading, we will be able to load specific parts of pretrained models separately.

## Problem Statement
It makes the project even more modular and convenient to use.

## Proposed Solution
Add options to choose which part of a saved model to load, without forgetting to check compatibility.

## Alternatives Considered
None.

## Implementation Plan

This feature can be broke down into 1 PR per training loop as they have different needs.

- [ ] Step 1: Ensure enough information is saved for modular loading, especially the environment name.
- [ ] Step 2: Breakdown model loading with choice by user.

The user will specify the parts to load with (in order of priority, a top loaded part can be overwritten by a lower one):
- `--load` for the entire model
- `--load-vision` for the vision part
- `--load-memory` for the memory part
- `--load-controller` for the controller part
- `--load-reward-predictor` for the reward predictor

## Additional Context
See https://docs.pytorch.org/tutorials/beginner/saving_loading_models.html and existing code about saving/loading.

## Acceptance Criteria
Safe (no code injection possible) saving/loading of models.

### LINKED ISSUE #58: [FEATURE] Ability to add pretrain part into the full training
## Summary
Ability to add pretrain part into the full training

## Problem Statement
Hard currently to start from a pretrained Vision model or VM model to train the entire World model agent

## Proposed Solution
How would you implement this feature? Include:
- Which components would be affected (Vision/Memory/Controller/Training)
- What changes would be needed
- Any architectural implications

## Alternatives Considered
What other approaches have you thought about?

## Implementation Plan
If this is a large feature, how should it be broken down into smaller PRs?

- [ ] Step 1: ...
- [ ] Step 2: ...
- [ ] Step 3: ...

## Additional Context
Any other information, examples, or references.

## Acceptance Criteria
How will we know this feature is complete and working?

- [ ] Criterion 1
- [ ] Criterion 2
- [ ] Tests added
- [ ] Documentation updated

### PR DESCRIPTION
**Pull request by Larwive**
Added a patch model which is used to load parts from it on top of the loaded model. Refs: #31, #58

# [GH-31] Allow modular loading of models

## Summary

New type of loading after previous full model loading. Patching allow loading specific parts of a pretrained model.

## Related Issue
Closes #31
Closes #58

## Changes Made

1. Add a new type of loading: patch loading

Model creation follows this order:
1. Creation from command line arguments.
2. If `--load-path` is specified, overwrite the created model with the one to load.
3. If `--patch-load-path` is specified, overwrite parts of the model depending on the value of `--patch`

## Type of Change

- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [ ] Documentation update
- [ ] Code refactoring
- [ ] Performance improvement
- [ ] Test addition/update

## Component(s) Affected

- [ ] Vision
- [ ] Memory
- [ ] Controller
- [x] WorldModel
- [ ] Training Scripts
- [ ] Tests
- [ ] Documentation
- [ ] Other: ___________

## Testing

- [ ] Unit tests added/updated
- [ ] Manual testing completed
- [ ] Tested on: [environment, e.g., CartPole-v1, CPU/GPU]

## Code Quality Checklist

- [ ] Code follows project style guidelines
- [ ] Docstrings added/updated
- [ ] Commit messages follow conventions

## Documentation

- [x] README.md updated
- [ ] Docstrings updated
- [ ] Comments added where logic is non-obvious

## Additional Notes

---

**Reviewer Checklist** (for maintainers):

- [ ] Code quality meets standards
- [ ] Tests are comprehensive
- [ ] Documentation is clear
- [ ] No unnecessary changes
- [ ] Commit messages are descriptive

### CODE REVIEW

**Comment by Bugz-gg** [comment]
Something is wrong in the cli I cannot find where 
`Available commands:
  - 0: Run
  - 1: Edit main arguments
  - 2: Edit advanced arguments
  - 3: See PPO arguments
  - 4: Switch to inference mode
  - 7: Save args
  - 8: Load saved args
  - exit: Exit`

**Comment by Bugz-gg** [comment]
I only use pretraied vision and in the patch there is vmc

**Comment by Larwive** [comment]
> Something is wrong in the cli I cannot find where `Available commands:
> 
> * 0: Run
> * 1: Edit main arguments
> * 2: Edit advanced arguments
> * 3: See PPO arguments
> * 4: Switch to inference mode
> * 7: Save args
> * 8: Load saved args
> * exit: Exit`

You mean the gap in the numbers ? Exiting used to be `9` and saving and loading are quite different in use compared to other commands. I still find the placement convenient on a numpad.

Do you have any proposition if you don't use a numpad ? It is possible to create aliases.

**Comment by Larwive** [comment]
> I only use pretraied vision and in the patch there is vmc

You need to specify `--patch v` if you want to only keep the vision from the model of `--patch-load-path`.
It has the same behavior than using `--load-path` if `--patch` is `vmc` except the loading is done after, so it overwrites.

I retested and can confirm it is working. Maybe the behavior is clearer if you see what I did to test.

First, I added at line 231 of `main.py`:
```python
logger.info(f"Model: {world_model.vision.__class_.__name__},  {world_model.memory.__class_.__name__},  {world_model.controller.__class_.__name__}")
exit(0)
```

I have a pretrained model with `VQ_VAE`, `LSTMMemory` and `DeepContinuousController`.
Running `python src/main.py --patch-load-path saved_models/ppo.pt` gives the output

`Model: VQ_VAE, LSTMMemory, DeepContinuousController`

Then, I'll add `--vision JEPA --memory TemporalTransformer --controller StochasticController` to the command line above.
Running as is still gives the same output because patching is done after model initialization with specified models: 
`Model: VQ_VAE, LSTMMemory, DeepContinuousController`

Then:
- `--patch v` will output `Model: *VQ_VAE*, TemporalTransformer, StochasticController`
- `--patch m` will output `Model: JEPA, *LSTMMemory*, StochasticController`
- `--patch c` will output `Model: JEPA, TemporalTransformer, *DeepContinuousController*`
- `--patch vm` will output `Model: *VQ_VAE*, *LSTMMemory*, StochasticController`
- `--patch vc` will output `Model: *VQ_VAE*, TemporalTransformer,  *DeepContinuousController*`
- `--patch mc` will output `Model: JEPA, *LSTMMemory*,  *DeepContinuousController*`
- `--patch vmc` will output `Model: *VQ_VAE*, *LSTMMemory*,  *DeepContinuousController*`

**Review by Bugz-gg — APPROVED** [review]
